README.md
# Time

A no-nonsense library for working with time and doing temporal calculations the easy way.

The aim of the **Time toolkit** is to help working with temporal data
like work timer measurements, schedules, time-tables, subscription timers etc.

For example, `Time` supports working with **negative times** and **times exceeding 24 hours**, as opposed to `DateTime` or `Carbon`:
```php
// negative time values
(string) Time::create('-10:30'); // -10:30:00
(string) Time::create('10:30')->subHours(100); // -89:30:00

// time values exceeding 24 hours
Time::create('50:00:00'); // 50 hours
Time::fromDays(10); // 10 days
```

On the contrary to DateTime or Carbon (nesbot/carbon), `Time` is simply a container for time measured in seconds,
it does not refer to any real point-in-time, there are no time zones no nothing.

With `Time` you can **compare** time values, perform **atirhmetic operations** or **convert** them.
You can **parse strings** containing formatted time values and you can print the `Time` object according to a specified **format**.

Your data may be stored in seconds, minutes, hours or in formatted strings, working with `Time` will still be the same.


## Methods

There are many methods for **handling**, **converting**, **modifying**, **reading** and **printing** the time value.

Options to create a time object:
```php
// constructor, create, fromSeconds, fromMinutes, fromHours, fromDays, fromWeeks
new Time('12:20') == Time::create('12:20')
Time::fromDays(4)
Time::fromHours(12, 20) == Time::create('12:20') == Time::create(44400) == Time::fromSeconds(44400)
```

Arithmetic methods:
```php
// arithmetic methods: add, sub, mult, div, mod
(new Time('12:00'))->add(Time::fromSeconds(17000))->sub(new Time(45))
Time::fromHours(10)->div(2) // 5 hours
```

Comparisons:
```php
// comparison methods: lt, lte, gt, gte, eq, neq, between
(new Time('12:00'))->between('12:30', '11:30') // TRUE
```

Adding / subtracting seconds, minutes, hours, days or weeks:
```php
// adding values: addSeconds, addMinutes, addHours, addDays, addWeeks
// subtracting values: subSeconds, subMinutes, subHours, subDays, subWeeks
(string) Time::create(123) //  00:02:03
        ->addSeconds(10)   //  00:02:13
        ->addSeconds(-10)  //  00:02:03
        ->subMinutes(-10)  // -00:08:03
```

Conversion options:
```php
// conversion options: toSeconds, toMinutes, toHours, toDays, toWeeks
Time::fromWeeks(4)->toSeconds() // 2419200
Time::fromDays(3.5)->toWeeks()  // 0.5
```

Input parsing (reading) and output formatting:
```php
// string time parsing (reading)
Time::create('23:59:59')->toSeconds() == TimeHelper::parse('23:59:59')
TimeHelper::parse('-10:30')
Time::create(TimeHelper::parse('10:30 PM', TimeHelper::FORMAT_HMA)) // custom format

// output formatting
(string) Time::create(123)->format(Time::FORMAT_HM); // 00:02 - custom format (HH:mm)
(string) Time::fromHours(2, 3, 4); // 02:03:04 - the default format (HH:mm:ss)
```

Output to other time objects (conversion):
```php
// converting to DateTime or Carbon: toDateTime, toCarbon
$carbon = Time::create(123)->toCarbon();
$datetm = Time::create('07:50 AM')->toDateTime();
```

Validating valid day time and clipping to valid day time:
```php
// clipping to valid day time
Time::create(-1)->isValidDayTime(); // FALSE
(string) Time::create(-1); // -00:00:01
(string) Time::create(-1)->clipToDayTime(); // 23:59:59
(Time::create('23:59:59'))          // 23:59:59
        ->addSeconds(1)             // 24:00:00
        ->clipToDayTime();          // 00:00:00
```

And there is more!

>**Note**: For all the methods, please refer to the **source code**.

## Mutable and Immutable Time objects

The default `Time` object is **immutable**.
It means that once a `Time` instance is created, its value does not change. Upon any modification a new instance is returned.
```php
$immutable = Time::fromSeconds(0);
// all the operations work as expected:
(string) $immutable->addSeconds(30)->mult(2); // "00:01:00"
// but the instance itself does not change - this is in contrast to the mutable TimeMutable object:
$immutable->getMinutes(); // 0
```
Sometimes one needs to treat a time object as a **mutable object**, an accumulator, the solution is the `TimeMutable` class.
```php
$mutable = TimeMutable::fromSeconds(0);
(string) $mutable->addSeconds(30)->mult(2); // "00:01:00"
// the modifications are accumulated inside the TimeMutable instance:
$mutable->getMinutes(); // 1
```
`TimeMutable` may be useful for aggregations:
```php
$acc = new TimeMutable();
foreach(... as $foo){
    $acc->add($foo->getDuration());
}
print $acc;
```

## Parsing time strings and formatting

By default, `Time` accepts time strings in `HH:MM:SS` format:
```php
Time::create('23:59:59')->sub('12:30')
```
The default format is stored in `TimeHelper::$defaultFormat` static variable and can be changed.
However, it is always more flexible to use a factory and pass `Time` instances into calculations:
```php
$timeFactory = new TimeFactory();
$timeFactory->create('23:59:59')->sub($timeFactory->create('12:30'))
```
This way the format of the input can be changed at runtime and can be set for all the factories independently.
The factories can also be implemented in custom way and use custom parsers.

To use **different formats** without using a factory, `TimeHelper` can be used manually:
```php
Time::create(TimeHelper::parse('23:59:59'))->sub(TimeHelper::parse('10:30 PM', TimeHelper::FORMAT_HMA))
```

The same principle is true for formatting. Converting the time objects to string is done using the format stored in `TimeHelper::$defaultFormat`, that is `HH:MM:SS` by default.

## Milliseconds, microseconds...

The `Time` object is built for calculations in **seconds**, which is the maximum resolution for which all the features work.

However, it is also possible to use time as double to get heigher resolution (milliseconds or microseconds):
```php
Time::create(5.500) // 5.5 seconds
```
Features that do not work correctly (yet) when using double:
- parsing - custom parsing needed
- formatting - `TimeHelper` is currently unable to format double correctly
- there is no round or abs implementation yet

>**Note**: When using **double** data type all the PHP implementation limitations apply.

## Installation
The easiest way to install Time is to use [Composer](https://getcomposer.org/). Just add `"dakujem/time"` to the "require" section in your `composer.json` file, like this:
```json
{
	"require": {
		"dakujem/time": '*'
	}
}
```

> **Note**: `dakujem/time` is built for and tested to run on PHP 5.6 and PHP 7 and above, however, it should also run on PHP 5.4 and 5.5 if needed.



changelog.md

# Changelog

> 📖 back to [readme](readme.md)

Time follows semantic versioning.\
Any issues should be reported.


## v1.1

Supports PHP 8.1 and rises minimum PHP version to 7.2.

**Edge case**\
Fixes previously incoherent behaviour of methods `Time::mod` and `Time::clipToDayTime` when the internal value is `null`.\
This has been fixed and calling these methods will result in a new `Time` instance with internal `null` value as well (previously it was `0`).


## v1

The initial stable version.


tests/index.php
<?php

declare(strict_types=1);

/**
 * This file is a part of dakujem/time package.
 * @author Andrej Rypak (dakujem) <xrypak@gmail.com>
 */
//
$dir = '.';

// run the test
$time1 = microtime(true);
require_once($dir . '/helper.phpt');
print '<hr/><pre>helper.phpt | Finished at: ' . date('Y-m-d H:i:s') . ' | Runtime: ' . (microtime(true) - $time1) . 's</pre>';

$time2 = microtime(true);
require_once($dir . '/time.phpt');
print '<hr/><pre>time.phpt | Finished at: ' . date('Y-m-d H:i:s') . ' | Runtime: ' . (microtime(true) - $time2) . 's</pre>';


tests/bootstrap.php
<?php

declare(strict_types=1);

/**
 * This file is a part of dakujem/time package.
 * @author Andrej Rypak (dakujem) <xrypak@gmail.com>
 */


namespace Dakujem\Time\Test;

use Tester\Assert;
use Tester\Environment;
use Tracy\Debugger;
use function getallheaders;

define('ROOT', __DIR__);

require_once __DIR__ . '/../vendor/autoload.php';
if (!class_exists(Assert::class)) {
    echo "Install Nette Tester using `composer update --dev`\n";
    exit(1);
}

// tester
Environment::setup();

// debugging
if (function_exists('getallheaders') && !empty(getallheaders()) && class_exists(Debugger::class)) {
    Debugger::$strictMode = true;
    Debugger::enable();
    Debugger::$maxDepth = 10;
    Debugger::$maxLen = 500;
}


// dump shortcut
function dump($var, $return = false)
{
    return Debugger::dump($var, $return);
}


src/Time.php
<?php

declare(strict_types=1);

namespace Dakujem;

use Carbon\Carbon;
use Carbon\CarbonImmutable;
use Carbon\CarbonInterface;
use DateTime;
use RuntimeException;

/**
 * Time object.
 *
 * Note:    internally, the time is kept in seconds, so the maximum supported integer resolution is one second.
 *            It is possible to use floating point numbers to keep fractions of seconds (milliseconds, microseconds, etc.).
 *
 *
 * @author Andrej Rypak (dakujem) <xrypak@gmail.com>
 */
class Time implements TimeInterface
{
    /**
     * @var int|double|NULL the time in seconds.
     */
    protected $time = null;

    public function __construct($time = null)
    {
        if ($time !== null) {
            $this->_override($this->parse($time));
        }
    }

    /**
     * Add given time value:
     * $this + $time
     *
     *
     * @param int|string|static|DateTime|CarbonInterface $time any parsable time format
     * @return static
     */
    public function add($time)
    {
        return $this->_set($this->_get() + $this->parse($time));
    }

    /**
     * Subtract given time value:
     * $this - $time
     *
     *
     * @param int|string|static|DateTime|Carbon $time any parsable time format
     * @return static
     */
    public function sub($time)
    {
        return $this->_set($this->_get() - $this->parse($time));
    }

    /**
     * Multiply the time by $x:
     * $this * $x
     *
     *
     * @param int|double $x
     * @return static
     */
    public function mult($x)
    {
        return $this->_set($this->_get() * $x);
    }

    /**
     * Divide the time by $x:
     * $this / $x
     *
     * Note: division by zero results in INF.
     *
     * @param int|double $x
     * @return int|double
     */
    public function div($x)
    {
        if (!is_int($x) && !is_float($x)) {
            $x = (double)$x;
        }
        return $this->_set($x !== 0.0 && $x !== 0 ? $this->_get() / $x : INF);
    }

    /**
     * Modulate the time by $x:
     * $this % $x
     *
     * Note: division by zero results in NAN.
     *
     * @param int|double $x
     * @return int|double
     */
    public function mod($x)
    {
        if (!is_int($x) && !is_float($x)) {
            $x = (double)$x;
        }
        if ($x === 0.0 || $x === 0) {
            return $this->_set(NAN);
        }
        if ($this->isNULL()) {
            return $this->_set(null);
        }
        return $this->_set(is_int($this->_get()) && is_int($x) ? $this->_get() % $x : fmod($this->_get(), $x));
    }

    /**
     * Perform a less-than comparison:
     * $this < $time
     *
     *
     * @param int|string|static|DateTime|CarbonInterface $time any parsable time format
     * @return bool
     */
    public function lt($time)
    {
        return $this->_get() < $this->parse($time);
    }

    /**
     * Perform a less-than-or-equal comparison:
     * $this <= $time
     *
     *
     * @param int|string|static|DateTime|CarbonInterface $time any parsable time format
     * @return bool
     */
    public function lte($time)
    {
        return $this->_get() <= $this->parse($time);
    }

    /**
     * Perform a greater-than comparison:
     * $this > $time
     *
     *
     * @param int|string|static|DateTime|CarbonInterface $time any parsable time format
     * @return bool
     */
    public function gt($time)
    {
        return $this->_get() > $this->parse($time);
    }

    /**
     * Perform a greater-than-or-equal comparison:
     * $this >= $time
     *
     *
     * @param int|string|static|DateTime|CarbonInterface $time any parsable time format
     * @return bool
     */
    public function gte($time)
    {
        return $this->_get() >= $this->parse($time);
    }

    /**
     * Perform a equal-to comparison:
     * $this == $time
     *
     *
     * @param int|string|static|DateTime|CarbonInterface $time any parsable time format
     * @return bool
     */
    public function eq($time)
    {
        return $this->_get() === $this->parse($time);
    }

    /**
     * Perform a not-equal-to comparison:
     * $this != $time
     *
     *
     * @param int|string|static|DateTime|CarbonInterface $time any parsable time format
     * @return bool
     */
    public function neq($time)
    {
        return $this->_get() !== $this->parse($time);
    }

    /**
     * Return TRUE when the time is within the interval defined by the given times.
     *
     * Performs one of these comparisons:
     *   $timeMin <= $this <= $timeMax    // this comparison is done for $sharp==FALSE (the default)
     *   $timeMin <  $this <  $timeMax    // this comparison is done for $sharp==TRUE
     *
     * Note: $timeMin and $timeMax are determined from $time1 and $time2.
     *
     *
     * @param int|string|static|DateTime|CarbonInterface $time1 any parsable time format
     * @param int|string|static|DateTime|CarbonInterface $time2 any parsable time format
     * @param bool $sharp [=FALSE] exclude the extremes of the interval?
     * @return bool
     */
    public function between($time1, $time2, $sharp = false)
    {
        $t1 = $this->parse($time1);
        $t2 = $this->parse($time2);
        if ($t1 <= $t2) {
            $from = $t1;
            $to = $t2;
        } else {
            $from = $t2;
            $to = $t1;
        }
        return
            $sharp ?
                $this->_get() > $from && $this->_get() < $to :
                $this->_get() >= $from && $this->_get() <= $to;
    }

    /**
     * Is the time a valid day time?
     * e.g. Is the time between 00:00:00 and 23:59:59 ?
     *
     *
     * @return bool
     */
    public function isValidDayTime()
    {
        return $this->_get() >= 0 && $this->_get() < self::DAY;
    }

    /**
     * Clip the time to a valid day time between 00:00:00 and 23:59:59.
     * This will perform a modulo-DAY operation: $this % DAY
     *
     *
     * @return static containing time between 00:00:00 and 23:59:59
     */
    public function clipToDayTime()
    {
        if ($this->isNULL()) {
            return $this->_set(null);
        }
        $t = is_int($this->_get()) ? $this->_get() % self::DAY : fmod($this->_get(), self::DAY);
        return $this->_set($t < 0 ? $t + self::DAY : $t);
    }

    public function addSeconds($seconds = 1)
    {
        return $this->_set($this->_get() + $seconds * self::SECOND);
    }

    public function addMinutes($minutes = 1)
    {
        return $this->_set($this->_get() + $minutes * self::MINUTE);
    }

    public function addHours($hours = 1)
    {
        return $this->_set($this->_get() + $hours * self::HOUR);
    }

    public function addDays($days = 1)
    {
        return $this->_set($this->_get() + $days * self::DAY);
    }

    public function addWeeks($weeks = 1)
    {
        return $this->_set($this->_get() + $weeks * self::WEEK);
    }

    public function subSeconds($seconds = 1)
    {
        return $this->addSeconds($seconds * -1);
    }

    public function subMinutes($minutes = 1)
    {
        return $this->addMinutes($minutes * -1);
    }

    public function subHours($hours = 1)
    {
        return $this->addHours($hours * -1);
    }

    public function subDays($days = 1)
    {
        return $this->addDays($days * -1);
    }

    public function subWeeks($weeks = 1)
    {
        return $this->addWeeks($weeks * -1);
    }

    /**
     * Returns the value of the internal time member.
     *
     * Warning:    this method is not intended to be used for retrieving time.
     *            It is provided for forward compatibility and testing purposes only.
     *            In case you need to check for NULL value, use $time->isNull() instead.
     *
     *
     * @return int|double|NULL
     */
    public function getRaw()
    {
        return $this->_get();
    }

    /**
     * Indicate whether the time is equal to zero.
     * @see isNULL()
     *
     *
     * @return bool
     */
    public function isZero()
    {
        return $this->_get() === 0;
    }

    /**
     * Indicate whether the time is NULL,
     * which means the time was not set or has been reset to NULL.
     * @see isZero()
     *
     *
     * @return bool
     */
    public function isNULL()
    {
        return $this->_get() === null;
    }

    /**
     * Indicate whether the time is negative or not.
     *
     *
     * @return bool TRUE for any negative value, FALSE for positive and zero time
     */
    public function isNegative()
    {
        return $this->_get() < 0;
    }

    /**
     * Returns -1 when the time is negative, 0 when it is zero, +1 when it is positive.
     *
     *
     * @return int
     */
    public function getSignum()
    {
        $s = $this->_get() * self::SECOND;
        return $s < 0 ? -1 : ($s === 0 ? 0 : 1);
    }

    /**
     * Return the hour part of the time.
     * WARNING: this does not return the time converted to hours! For that purpose, use the toHours() method.
     *
     *  HH:MM:SS
     *  \/
     *
     * @return int
     */
    public function getHours()
    {
        return (int)abs((int)($this->_get() / self::HOUR));
    }

    /**
     * Return the minute part of the time.
     * WARNING: this does not return the time converted to minutes! For that purpose, use the toMinutes() method.
     *
     *  HH:MM:SS
     *     \/
     *
     * @return int
     */
    public function getMinutes()
    {
        return (int)abs(((int)($this->_get() / self::MINUTE)) % self::MINUTE);
    }

    /**
     * Return the seconds part of the time.
     * WARNING: this does not return the time converted to seconds! For that purpose, use the toSeconds() method.
     *
     *  HH:MM:SS.frac
     *        \/
     *
     * @return int
     */
    public function getSeconds()
    {
        return (int)abs((int)$this->_get() % self::MINUTE);
    }

    /**
     * Return the remaining fraction of a second. Returns NULL when the time value is integer.
     *
     *  HH:MM:SS.frac
     *           \__/
     *
     * @return double|NULL
     */
    public function getSecondFraction()
    {
        $val = $this->_get();
        return is_float($val) ? abs(fmod($val, self::SECOND)) : null;
    }

    /**
     * Return the stored time in seconds.
     *
     *
     * @return int|double
     */
    public function toSeconds()
    {
        return $this->_get() * self::SECOND;
    }

    /**
     * Return the stored time in minutes.
     *
     *
     * @return int|double
     */
    public function toMinutes()
    {
        return $this->_get() / self::MINUTE;
    }

    /**
     * Return the stored time in hours.
     *
     *
     * @return int|double
     */
    public function toHours()
    {
        return $this->_get() / self::HOUR;
    }

    /**
     * Return the stored time in days.
     *
     *
     * @return int|double
     */
    public function toDays()
    {
        return $this->_get() / self::DAY;
    }

    /**
     * Return the stored time in weeks.
     *
     *
     * @return int|double
     */
    public function toWeeks()
    {
        return $this->_get() / self::WEEK;
    }

    /**
     * Create and return a DateTime instance using "H:i:s" format.
     * Note: this will clip the stored time to a valid day time value using clipToDayTime() method.
     *
     *
     * @return DateTime
     */
    public function toDateTime()
    {
        return new DateTime((string)$this->clipToDayTime());
    }

    /**
     * Fill a Carbon instance using Carbon::hour(), Carbon::minute() and Carbon::second() methods.
     * If a Carbon instance is not provided, a new one will be created.
     *
     *
     * @param CarbonInterface|null $target an instance to fill
     * @return CarbonInterface
     */
    public function toCarbon(CarbonInterface $target = null, bool $immutable = true) //: CarbonInterface
    {
        if ($target === null) {
            $target = $immutable ? new CarbonImmutable() : new Carbon();
        }
        return $target->hour($this->getHours())->minute($this->getMinutes())->second($this->getSeconds());
    }

    /**
     * Create and return a copy of self.
     *
     *
     * @return static a copy of the original Time object, for fluent calls
     */
    public function copy()
    {
        return clone $this;
    }

    /**
     * Returns the time in seconds or NULL.
     *
     *
     * @param mixed $time
     * @return int|NULL returns NULL when the time passed is NULL or an empty string
     * @throws RuntimeException
     */
    protected function parse($time)
    {
        if ($time instanceof self) {
            return $time->_get();
        }
        return TimeHelper::parse($time);
    }

    /**
     * Format the time using the format provided.
     *
     * The Time::FORMAT_* constants can be used to format the time to most used time formats.
     * Recognized characters for format are "?+HhisGgAa".
     * - see php date() function for "HhisGgAa",
     * - '?' is used for the minus sign, only present when the time is negative
     * - '+' is used for minus and plus sign, always present
     *
     *
     * @param string|NULL $format
     * @return string formatted time
     */
    public function format($format = null)
    {
        return TimeHelper::format($this, $format);
    }

    public function __toString()
    {
        return $this->format(null);
    }

    /**
     * Internal setter.
     * @internal
     */
    protected function _set($value)
    {
        $mutation = $this->copy();
        $mutation->_override($value);
        return $mutation;
    }

    /**
     * Internal getter.
     * @internal
     */
    protected function _get()
    {
        return $this->time;
    }

    /**
     * This method is only intended to be called from the constructor or directly after clonning.
     *
     *
     * @param int|double|NULL $value
     */
    private function _override($value)
    {
        $this->time = $value === null ? null : $value * self::SECOND;
    }

    /**
     * Universal static factory.
     *
     *
     * @param mixed $time any parsable time or NULL
     * @return static the Time object
     */
    public static function create($time = null)
    {
        return new static($time);
    }

    public static function fromSeconds($seconds)
    {
        return static::create(self::calculateSeconds(0, 0, 0, 0, $seconds));
    }

    public static function fromMinutes($minutes, $seconds = 0)
    {
        return static::create(self::calculateSeconds(0, 0, 0, $minutes, $seconds));
    }

    public static function fromHours($hours, $minutes = 0, $seconds = 0)
    {
        return static::create(self::calculateSeconds(0, 0, $hours, $minutes, $seconds));
    }

    public static function fromDays($days, $hours = 0, $minutes = 0, $seconds = 0)
    {
        return static::create(self::calculateSeconds(0, $days, $hours, $minutes, $seconds));
    }

    public static function fromWeeks($weeks, $days = 0, $hours = 0, $minutes = 0, $seconds = 0)
    {
        return static::create(self::calculateSeconds($weeks, $days, $hours, $minutes, $seconds));
    }

    private static function calculateSeconds($weeks, $days, $hours, $minutes, $seconds)
    {
        return
            (is_numeric($weeks) ? $weeks : (int)$weeks) * self::WEEK +
            (is_numeric($days) ? $days : (int)$days) * self::DAY +
            (is_numeric($hours) ? $hours : (int)$hours) * self::HOUR +
            (is_numeric($minutes) ? $minutes : (int)$minutes) * self::MINUTE +
            (is_numeric($seconds) ? $seconds : (int)$seconds) * self::SECOND;
    }
}


src/TimeFactory.php
<?php

declare(strict_types=1);

namespace Dakujem;

/**
 * TimeFactory - a factory service.
 *
 * It can be used to create Time or TimeMutable objects with parsing input beforehand.
 * Its advantage for users is the ability to specify the input format (for parsing strings).
 * The factory can also be used when static factory calls of the time objects in code are not desired.
 *
 * TimeFactory::$immutable flag determins, which of the two time objects will be created - Time (immutable) or TimeMutable.
 * TimeFactory::$format will be passed to the TimeHelper::parse($rawTime, $format) call.
 *
 *
 * @author Andrej Rypak <xrypak@gmail.com>
 */
final class TimeFactory
{
    private $format = TimeHelper::FORMAT_HMS;
    private $immutable = true;

    public function __construct($format = null, $immutable = null)
    {
        $format !== null && $this->setFormat($format);
        $immutable !== null && $this->setImmutable($immutable);
    }

    public function create($rawTime = null, $format = null)
    {
        if ($format === null) {
            $format = $this->getFormat();
        }
        $seconds = TimeHelper::parse($rawTime, $format);
        return $this->getImmutable() ? new Time($seconds) : new TimeMutable($seconds);
    }

    public function getFormat()
    {
        return $this->format;
    }

    public function setFormat($format)
    {
        $this->format = $format;
        return $this;
    }

    public function useFormatHoursMinutes()
    {
        return $this->setFormat(Time::FORMAT_HM);
    }

    public function useFormatHoursMinutesSeconds()
    {
        return $this->setFormat(Time::FORMAT_HMS);
    }

    public function getImmutable()
    {
        return $this->immutable;
    }

    public function setImmutable($immutable)
    {
        $this->immutable = !!$immutable;
        return $this;
    }
}


src/TimeMutable.php
<?php

declare(strict_types=1);

namespace Dakujem;

/**
 * A mutable time object.
 *
 * All calls that modify the time value are accumulated instead of returning a new instance with the new value.
 * Also provides the method set().
 *
 *
 * @author Andrej Rypak (dakujem) <xrypak@gmail.com>
 */
class TimeMutable extends Time
{
    /**
     * Set the time.
     *
     *
     * @param int|string|static|DateTime|Carbon $time any parsable time format
     * @return static
     */
    public function set($time)
    {
        return $this->_set($this->parse($time));
    }

    /**
     * Internal setter.
     * @note: does NOT produce a copy of self
     * @internal
     */
    protected function _set($value)
    {
        $this->time = $value === null ? null : $value * self::SECOND;
        return $this;
    }
}


src/TimeHelper.php
<?php

declare(strict_types=1);

namespace Dakujem;

use Carbon\Carbon;
use DateTime;
use RuntimeException;

/**
 * TimeHelper.
 *
 * Note:    only time formats defined in constants are supported officially,
 *            however, the use of custom formats is possible.
 *            Avoiding funny formats is advisable though.
 *            Known not to work (parsing): "?i:s"
 *
 *
 * @author Andrej Rypak (dakujem) <xrypak@gmail.com>
 */
class TimeHelper
{
    //
    // format consts
    //
    const FORMAT_HMS = '?H:i:s'; //           02:34:56     or  -123:45:59
    const FORMAT_HM = '?H:i'; //              02:34        or  -123:45
    const FORMAT_HMS_SIGNED = '+H:i:s'; //   +02:34:56     or  -123:45:59
    const FORMAT_HM_SIGNED = '+H:i'; //      +02:34        or  -123:45
    const FORMAT_HMSA = 'h:i:s A'; //         12:34:56 AM  or    01:23:45 PM
    const FORMAT_HMA = 'h:i A'; //            12:34 AM     or    01:23 PM

    /**
     * Default format (static).
     * @var string
     */
    public static $defaultFormat = self::FORMAT_HMS;

    /**
     * Format the time using the format provided.
     *
     * The Time::FORMAT_* constants can be used to format the time to most used time formats.
     * Recognized characters for format are "?+HhisGgAa".
     * - see php date() function for "HhisGgAa",
     * - '?' is used for the minus sign, only present when the time is negative
     * - '+' is used for minus and plus sign, always present
     *
     *
     * @param TimeInterface $time
     * @param string|NULL $format
     * @return string formatted time
     */
    public static function format(TimeInterface $time, $format = null)
    {
        $source = $format !== null && $format !== '' ? $format : self::$defaultFormat;
        $neg = $time->isNegative();
        $v = $time->isValidDayTime();
        $h = $time->getHours();
        $h12 = $v ? ($h % 12 === 0 ? 12 : $h % 12) : $h;
        $m = $time->getMinutes();
        $s = $time->getSeconds();
        return str_replace(['?', '+', 'H', 'h', 'G', 'g', 'i', 's', 'A', 'a'], [
            $neg ? '-' : '', // ?
            $neg ? '-' : '+', // +
            sprintf('%02d', $h), // H
            sprintf('%02d', $h12), // h
            $h, // G
            $h12, // g
            sprintf('%02d', $m), // i
            sprintf('%02d', $s), // s
            $v ? ($h < 12 ? 'AM' : 'PM') : '', // A
            $v ? ($h < 12 ? 'am' : 'pm') : '', // a
        ], $source);
    }

    /**
     * Returns the time in seconds or NULL.
     *
     *
     * @param mixed $time
     * @param string|NULL $format
     * @return int|NULL returns NULL when the time passed is NULL or an empty string
     * @throws RuntimeException
     */
    public static function parse($time, $format = null)
    {
        if (is_numeric($time)) {
            // regard it as seconds
            return $time;
        } elseif ($time === null || $time === '') {
            return null;
        } elseif (is_string($time)) {
            // regard it as time string
            return self::parseFormat($time, $format === null ? self::$defaultFormat : $format);
        } elseif (is_array($time)) {
            // [s, m, h, d, w]
            $s = reset($time);
            $m = next($time);
            $h = next($time);
            $d = next($time);
            $w = next($time);
            return empty($s) ? null : self::calculateSeconds($w, $d, $h, $m, $s);
        } elseif ($time instanceof TimeInterface) {
            return $time->toSeconds();
        } elseif ($time instanceof Carbon) {
            return self::parse([$time->second, $time->minute, $time->hour]);
        } elseif ($time instanceof DateTime) {
            return self::parse($time->format('H:i:s'));
        }
        throw new RuntimeException('Invalid argument passed.');
    }

    /**
     * @note: sign handling should be improved
     */
    private static function parseFormat($value, $format)
    {
        // 1/ -----------------------------------------------------------------------------------------
        // read the numbers form the input string

        $numbers = null;
        if (!preg_match('#([+-]?[0-9]+)(.([+-]?[0-9]+)?(.([+-]?[0-9]+)?)?)?#', $value, $numbers)) { // PREG_OFFSET_CAPTURE
            return null;
        }
        $hi = 1;
        $mi = 3;
        $si = 5;
        // $vals contain the first, second and third number found in the $value string
        $vals = [
            isset($numbers[$hi]) ? $numbers[$hi] : 0,
            isset($numbers[$mi]) ? $numbers[$mi] : 0,
            isset($numbers[$si]) ? $numbers[$si] : 0,
        ];

        // 2/ -----------------------------------------------------------------------------------------
        // according to the format string, decide which numbers denote hours, minutes and seconds

        $hpos1 = stripos($format, 'h');
        $hpos = $hpos1 !== false ? $hpos1 : stripos($format, 'g');
        $ipos = stripos($format, 'i');
        $spos = stripos($format, 's');
        // hpos, ipos, spos contain the position in $format
        if ($hpos === false && $ipos === false && $spos === false) {
            return null;
        }
        // $keys contain valid references to hours, minutes and seconds
        $h = $m = $s = 0;
        $keys = [];
        if ($hpos !== false) {
            $keys[$hpos] = &$h; // reference to hours
        }
        if ($ipos !== false) {
            $keys[$ipos] = &$m; // reference to minutes
        }
        if ($spos !== false) {
            $keys[$spos] = &$s; // reference to seconds
        }
        ksort($keys); // sort $keys according to occurence in $format
        foreach ($keys as &$ref) {
            // match the references in $keys with the values
            $ref = current($vals); // $vals contain values read from $value string
            next($vals);
        }

        // 3/ -----------------------------------------------------------------------------------------
        // correct negative values

        $hneg = substr((string)$h, 0, 1) === '-'; // hours negative
        $mneg = substr((string)$m, 0, 1) === '-'; // minutes negative
        $sneg = substr((string)$s, 0, 1) === '-'; // seconds negative
        if (true) {
            $h = (int)$h;
            $m = (int)$m;
            $s = (int)$s;
        }
        if (substr_count($format, '?') <= 1 && substr_count($format, '+') <= 1) {
            // this corrects the reading of times like -12:30, when format contains only one sign,
            // consequently, -12:30 will result in -12 hours and -30 minutes time, when format is ?H:i
            // when format is set as ?H:?i, this will not happen, and will result in time -11 hours and 30 minutes (-12 hours +30 minutes))
            if ($hneg) {
                $m = $mneg ? $m : -$m;
                $s = $sneg ? $s : -$s;
            } else {
                $m = !$mneg ? $m : -$m;
                $s = !$sneg ? $s : -$s;
            }
            // NOTE: this does not cover the case when format "?i:s" is used
        }

        // 4/ -----------------------------------------------------------------------------------------
        // check for and correct the 12-hour format (if used)

        $f12h = stripos($format, 'a') !== false; // check for 12-h format?
        if ($f12h) {
            if ($h > 12 || $h < 0) { // invalid 12h format
                return null;
            }
            $a = stripos($value, 'am');
            if ($a === false) {
                $p = stripos($value, 'pm');
                if ($p !== false) {
                    $am = false;
                } else {
                    return null; // AM or PM not found
                }
            } else {
                $am = true;
            }
            // now $am contains am/pm, correct the time
            if ($h == 12 && $am) {
                $h = 0;
            } elseif ($h != 12 && !$am) {
                $h = $h + 12; // PM
            }
        }

        // 5/ -----------------------------------------------------------------------------------------
        // return the result

        return self::calculateSeconds(0, 0, $h, $m, $s);
    }

    private static function calculateSeconds($weeks, $days, $hours, $minutes, $seconds)
    {
        return
            (is_numeric($weeks) ? $weeks : (int)$weeks) * TimeInterface::WEEK +
            (is_numeric($days) ? $days : (int)$days) * TimeInterface::DAY +
            (is_numeric($hours) ? $hours : (int)$hours) * TimeInterface::HOUR +
            (is_numeric($minutes) ? $minutes : (int)$minutes) * TimeInterface::MINUTE +
            (is_numeric($seconds) ? $seconds : (int)$seconds) * TimeInterface::SECOND;
    }
}
